monai/metrics/metric.py

```diff
@@ -0,0 +1,216 @@
+        """
+        computation logic for the y_pred and y of a iteration, the data should be `batch-first` Tensors.
+        Every subclass metric should implement its own computation logic according to its algorithm.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+
+class Cumulative:
+    """
+    Utility class for the typical cumulative computation process based on PyTorch Tensors.
+    It cumulates tensors in the buffer, then sync across distributed ranks and aggregate.
+    Note: the data list should have the same length every time calling `add()` in a round,
+    it will automatically create buffers according to the length of data list.
+    Execute the steps referring to below examples::
+
+        cum = Cumulative()
+        cum.add(x, y)
+        cum.add(a, b)
+        cum.add(c, d)
+        cum.agrregate()
+        result = cum.get_buffer()
+        cum.reset()
+
+    """
+
+    def __init__(self):
+        self.buffer_num: int = 0
+        self._buffers: Optional[List[List[torch.Tensor]]] = None
+        self._synced_tensors: Optional[List[Optional[torch.Tensor]]] = None
+        self._synced: bool = False
+
+    def reset(self):
+        """
+        Reset the buffers for cumulative tensors and the synced results.
+
+        """
+        self._buffers = None
+        self._synced_tensors = None
+        self._synced = False
+
+    def add(self, *data: torch.Tensor):
+        """
+        Add samples to the cumulative buffers.
+
+        Args:
+            data: list of input tensor, make sure the input data order is always the same in a round.
+                every item of data will be added to the corresponding buffer.
+
+        """
+        data_len = len(data)
+        if self._buffers is None:
+            self._buffers = [[] for _ in range(data_len)]
+        elif len(self._buffers) != data_len:
+            raise ValueError(f"data length: {data_len} doesn't match buffers length: {len(self._buffers)}.")
+        if self._synced_tensors is None:
+            self._synced_tensors = [None for _ in range(data_len)]
+
+        for i, d in enumerate(data):
+            if not isinstance(d, torch.Tensor):
+                raise ValueError(f"the data to cumulate in a buffer must be PyTorch Tensor, but got: {type(d)}.")
+            self._buffers[i].append(d)
+        self._synced = False
+
+    def aggregate(self):
+        """
+        Sync the buffers across distributed ranks and aggregate final results based on the buffers.
+        This base class only syncs the data, subclasses should implement more logic.
+
+        """
+        self._sync()
+
+    def _sync(self):
+        """
+        All gather the buffers across distributed ranks for aggregating.
+        Every buffer will be concatenated as a PyTorch Tensor.
+
+        """
+        self._synced_tensors = [evenly_divisible_all_gather(torch.cat(b, dim=0), concat=True) for b in self._buffers]
```

Comment:
may need to check `torch.distributed.is_available()` and `torch.distributed.is_initialized()`. Is this interface also supposed to be used for the case of single machine, multiprocessing? if not, how to do multiprocessing metrics computations? (for the example in line 112-118, `cum.add` may run in parallel to speed up the metrics computation)

Reply:
Hi @wyli ,

1. We already checked distributed status in `evently_divisible_all_gather()`:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/misc.py#L393
2. I added `torch.distributed.is_available()` check in the latest commit.
3. This interface is designed for `multi-processing` with the `sync` operation across processes, to speed up metrics computation, we usually split data to ranks by `DistributedSampler` before an epoch, every process then run its own data part according to the example in line 112 - 118. everytime we call `add()` in a process, it's a iteration, I think no need to call `add()` in parallel again, you can increase the batch size if you feel GPU is not full. I enhanced the doc-string of this interface for `multi-processing` in the latest commit.

Thanks.

Reply:
hi @Nic-Ma I'm not talking about training here (so not quite related to 'epoch' or 'batch'). For example, if we have 1000 model predictions already stored in nifti files, and 1000 corresponding ground truth files, how to efficiently compute 1000 dice scores and return an mean and std over the 1000 scores? the 1000 dice computations are independent to each other, a evaluation suite would use multiple processes to speed up the whole computation. my question is how to do it with these APIs?

Reply:
looks nice, could you please make a tutorial for this?

Reply:
Sure, I can make a tutorial for it when this PR merged.

Thanks.